### DISCUSSION TITLE
Drag & Drop devices in racks

### DISCUSSION THREAD

**Discussion by IcyPunch** [netbox-community/netbox#8740] [Ideas]
Is there a way to drag and drop devices in a rack? I used the search and looked for already created discussions but could not find anything.

In the past I resorted to doing everything manually. So I had to edit each device in a rack to reflect the physical change I made, which is very tedious. I did it via GUI. Export and re-import is in my opinion still a bit more tedious than it needs to do and often times not worth the effort.

**Comment by ziggekatten** [comment]
A small python script using pynetbox library where you have three command line params: device, desired rack and desired position. 

That would be my solution instead of using the GUI.

**Reply by jcollie** [reply]
Most of the people that will be doing day-to-day management of a data center won't have the skills necessary to implement a script, or the patience to use it once implemented.

**Reply by WagoL** [reply]
Does somebody has a script like this available?

**Reply by candlerb** [reply]
Since [v3.4.0](https://github.com/netbox-community/netbox/releases/v3.4.0), CSV bulk uploads are able to modify existing records as well as create new ones. Hence you should be able to do it with a CSV upload: `id,rack,position,face`

(In some cases you may also need to provide `site` and/or `location`, since rack names are not necessarily unique without them)

**Comment by IcyPunch** [comment]
that's not really what i'm looking for. It's also too complicated for normal users just clicking through the UI.
I would love the possibility to adjust devices placements within racks through the UI. Not editing each device manually and changing it's placement.
Just drag and drop in the rack overview would be perfect.

Do I need to create a feature request somewhere? where?

**Reply by jeremystretch** [reply]
This has already been proposed under #6726 and rejected due to the amount of work it would require. If you need this functionality, I strongly encourage you to implement it as a custom plugin for NetBox.

**Reply by IcyPunch** [reply]
That is very disheartening to be honest. I want to reduce the complexity for users.
I'm not a programmer at all so I can't provide an implentation of this myself :(

**Reply by rodvand** [reply]
I'm curious of your workflow and how often you move devices in your racks?

**Reply by IcyPunch** [reply]
It's not just about me. It's more so about the people I work with. They can't understand why something "simple" like this is not implemented as of right now.
We move stuff around often enough that it's a big pain in the butt to basically manually edit each device's location even if we just move one device technically for example.
I can't imagine we are the only ones that would like to have this feature. We don't do everything via python scripts, automation tools and stuff like that. That's too much to just be able to move stuff easily within a rack.

**Reply by peterbaumert** [reply]
> It's not just about me. It's more so about the people I work with. They can't understand why something "simple" like this is not implemented as of right now.
> 
> We move stuff around often enough that it's a big pain in the butt to basically manually edit each device's location even if we just move one device technically for example.
> 
> I can't imagine we are the only ones that would like to have this feature. We don't do everything via python scripts, automation tools and stuff like that. That's too much to just be able to move stuff easily within a rack.

Especially while planing and "somwthing comes up" and you have to rearrange a rack. Just happened to me and its annoying work to move many devices. I would really love this feature too.

**Reply by douginoz** [reply]
I'm at the stage of planning the rack layout.  That means rearranging devices many, many times as I add, remove, and change what goes where.  Visual feedback of the layout is important in being able to decide on best placement.  Having to manually do it by working out which slot(s) a device would move to, and what other devices would need to be shuffled around to accomodate it, is a nightmare and means I immediately don't wish to use netbox any more.  I'm sure that doesn't matter to the devs, but it should!

**Reply by kkthxbye-code** [reply]
> I'm sure that doesn't matter to the devs, but it should!

It doesn't have to matter to the devs, it's very possible to implement this as a plugin. As Jeremy mentions, if a good plugin PoC comes along, we might even be able to onboard this into the core. It's just a complex feature, there's no conspiracy here, no ones trying to make it harder for you to use netbox.

**Reply by candlerb** [reply]
> I'm at the stage of planning the rack layout. That means rearranging devices many, many times as I add, remove, and change what goes where. Visual feedback of the layout is important in being able to decide on best placement. Having to manually do it by working out which slot(s) a device would move to, and what other devices would need to be shuffled around to accomodate it, is a nightmare

I did a big data centre move about 3 years ago, and even though we were already using Netbox, I did all the planning for the new rack layouts in spreadsheets.  This was for two reasons: one was because Netbox didn't support easy drag and drop, but more importantly, the devices were already in Netbox, and Netbox can't store "planned" and "current" locations for the same device at the same time.

So I waited until each device had actually been moved into its new physical location, before updating it in Netbox.  An even bigger chunk of work was deleting and recreating all the cables: interfaces to new switches or new patch panels, power cables to new PDUs, console cables to new console servers etc.

**Reply by noc-cgesd** [reply]
We would love this feature. We are also planning some rack upgrades and this feature would be so helpful.

**Reply by NaClnPepper** [reply]
I spent several months migrating our datacenter documentation to Netbox when I had the time and this feature would be incredible. The ability to drag/drop the dashboards in 3.5.X is similar except it would take place in racks. However, I'm not a programmer either but I understand the complexity behind implementing something that seems simple on the surface.

The new device/rack view and next/back buttons have saved a lot of time anyway - very happy with those.

**Reply by candlerb** [reply]
Being able to show two devices side by side (with their interfaces), and drag-and-place cables between the two, would be even more awesome!

**Comment by julianze** [comment]
I just thought about an intermediate step.
When editing a device it could bei useful if there is the rack displayed side-by-side to the forms.
When clicking on a height unit in the rack, the form should be filled with the selected height unit.

This would Speed up mounting devices in the right position.
Often i have to correct the height, because i did not match the right position. This ends up in opening a second browser Tab to switch between two windows to choose the right position.

Clearly a problem is If only the selected rack in the forms should be visible or maybe buttons like "next rack" could help switching between neighbored racks visually.
